### PR TITLE
Jinja: Add support for Block Assignments

### LINKED ISSUE #2905: sqlfluff doesn't recognise a jinja variable set inside of "if" statement
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When I try to define a jinja variable using "set" jinja directive inside of an "if" jinja statement, sqlfluff complains: 
"Undefined jinja template variable".

### Expected Behaviour

to not have a linting issue

### Observed Behaviour

sqlfluff lint gives an error:
"Undefined jinja template variable"

### How to reproduce

try to create a "temp.sql" file with the following content

```
{% if True %}
    {% set some_var %}1{% endset %}
    SELECT {{some_var}}
{% endif %}
```

and run:
```
sqlfluff lint ./temp.sql
```

You will get the following error:
```
== [./temp.sql] FAIL                                                                                                                    
L:   2 | P:  12 |  TMP | Undefined jinja template variable: 'some_var'
L:   3 | P:  14 |  TMP | Undefined jinja template variable: 'some_var'
```

### Dialect

tested on 'snowflake' dialect

### Version

sqlfluff, version 0.11.1
Python 3.8.12

### Configuration

[sqlfluff]
verbose = 1
dialect = snowflake
templater = jinja
exclude_rules = L027,L031,L032,L036,L044,L046,L034,L050
output_line_length = 121
sql_file_exts=.sql

[sqlfluff:rules]
tab_space_size = 4
max_line_length = 250
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = aliases

[sqlfluff:rules:L042]
forbid_subquery_in = both

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = upper

[sqlfluff:rules:L014]
extended_capitalisation_policy = lower

[sqlfluff:rules:L030]  # function names
extended_capitalisation_policy = upper

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #2905

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/templaters/jinja.py [hunk 1]
```diff
@@ -343,20 +343,31 @@ def process(
         # first Exception which serves only to catch catastrophic errors.
         try:
             syntax_tree = env.parse(in_str)
-            undefined_variables = meta.find_undeclared_variables(syntax_tree)
+            maybe_undefined = meta.find_undeclared_variables(syntax_tree)
```
Comment (tunetheweb) [inline src/sqlfluff/core/templaters/jinja.py:346]: Not loving this name. WDYT about `potentially_undefined_variables`? Too long?

File: src/sqlfluff/core/templaters/jinja.py [hunk 2]
```diff
@@ -343,20 +343,31 @@ def process(
...
         except Exception as err:  # pragma: no cover
             # TODO: Add a url here so people can get more help.
             raise SQLTemplaterError(f"Failure in identifying Jinja variables: {err}.")
 
-        # Get rid of any that *are* actually defined.
-        for val in live_context:
-            if val in undefined_variables:
-                undefined_variables.remove(val)
+        really_undefined = set()
```
Comment (tunetheweb) [inline src/sqlfluff/core/templaters/jinja.py:351]: Similar here. Should we use the original `undefined_variables` name here?

File: test/core/templaters/jinja_test.py [hunk 3]
```diff
@@ -411,6 +411,20 @@ def test__templater_jinja_error_variable():
...
+    """Test no templater violation for variable defined within template."""
+    t = JinjaTemplater(override_context=dict(blah="foo"))
+    instr = """{% if True %}
+    {% set some_var %}1{% endset %}
+    SELECT {{some_var}}
+{% endif %}
+"""
+    outstr, vs = t.process(in_str=instr, fname="test", config=FluffConfig())
+    assert str(outstr) == "\n    \n    SELECT 1\n\n"
+    # Check we have violations.
```
Comment (tunetheweb) [inline test/core/templaters/jinja_test.py:424]:
```suggestion
    # Check we have no violations.
```

File: test/core/templaters/jinja_test.py [hunk 4]
```diff
@@ -411,6 +411,20 @@ def test__templater_jinja_error_variable():
...
+    assert not len(vs) > 0
```
Comment (tunetheweb) [inline test/core/templaters/jinja_test.py:425]: Why not this?

```suggestion
    assert len(vs) == 0
```

**Review by tunetheweb — CHANGES_REQUESTED** [review]

File: src/sqlfluff/core/templaters/jinja.py (hunk 1 above)
Comment (barrywhart) [reply src/sqlfluff/core/templaters/jinja.py:346]: Works for me!

File: test/core/templaters/jinja_test.py (hunk 3 above)
Comment (barrywhart) [reply test/core/templaters/jinja_test.py:424]: Heh, sloppy copy paste by me.

File: test/core/templaters/jinja_test.py (hunk 4 above)
Comment (barrywhart) [reply test/core/templaters/jinja_test.py:425]: Oh yes, stupid mistake on my part -- I was in a hurry and not paying attention. Will change.

**Comment by barrywhart** [comment]
@tunetheweb: Updated!

**Review by tunetheweb — APPROVED** [review]
LGTM.

Maybe change name of the PR to:

Jinja: Add support for Block Assignments

**Pull request merged**